icu4j/main/core/src/main/java/com/ibm/icu/text/DateTimePatternGenerator.java

```diff
@@ -173,15 +167,28 @@ private void initData(ULocale uLocale, boolean skipStdPatterns) {
 
     private void addICUPatterns(PatternInfo returnInfo, ULocale uLocale) {
         // first load with the ICU patterns
-        for (int i = DateFormat.FULL; i <= DateFormat.SHORT; ++i) {
-            SimpleDateFormat df = (SimpleDateFormat) DateFormat.getDateInstance(i, uLocale);
-            addPattern(df.toPattern(), false, returnInfo);
-            df = (SimpleDateFormat) DateFormat.getTimeInstance(i, uLocale);
-            addPattern(df.toPattern(), false, returnInfo);
-
-            if (i == DateFormat.SHORT) {
-                consumeShortTimePattern(df.toPattern(), returnInfo);
+        ICUResourceBundle rb = (ICUResourceBundle) UResourceBundle.getBundleInstance(ICUData.ICU_BASE_NAME, uLocale);
+        String calendarTypeToUse = getCalendarTypeToUse(uLocale);
+        // TODO: this is a HACK to work around ICU-22668/CLDR-17892!  Remove once they're fixed!
+        if (calendarTypeToUse.equals("iso8601")) {
+            calendarTypeToUse = "gregorian";
+        }
+        ICUResourceBundle dateTimePatterns = rb.getWithFallback("calendar/" + calendarTypeToUse + "/DateTimePatterns");
+        if (dateTimePatterns.getSize() < 8 || dateTimePatterns.getType() != UResourceBundle.ARRAY) {
```

Comment:
Shouldn't these two checks be reversed, first verifying that it is the expected type and then verifying the size of it? If it's an unknown type, it's unknown what the size means, isn't it?

Reply:
Done.